### PR TITLE
Fixed #32519 -- Added support for using key and path transforms in update() for JSONFields.

### PR DESCRIPTION
**Pull request by adzshaf**
# Trac ticket number

ticket-32519

# Branch description

This PR introduces support for ``JSONField`` key transforms in ``update`` so we can do partial update ``JSONField`` more efficiently. Consider following example to update a ``JSONField``.

```pycon
>>> user_preferences = UserPreferences.objects.create(
...     settings={
...         "theme": "dark",
...         "notifications": True,
...         "font": {"size": 10, "name": "Arial"},
...     }
... )
>>> UserPreferences.objects.update(
...     settings__theme="light",
...     settings__font__size=20,
... )
>>> user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
>>> print(user_preferences.settings)
{'theme': 'light', 'notifications': True, 'font': {'size': 20, 'name': 'Arial'}}
```

You can also remove a key by providing ``NOT_PROVIDED`` to parameter value.

```pycon
>>> from django.db.models import NOT_PROVIDED
>>> user_preferences = UserPreferences.objects.create(
...     settings={
...         "theme": "dark",
...         "notifications": True,
...         "font": {"size": 10, "name": "Arial"},
...     }
... )
>>> UserPreferences.objects.update(
...     settings__theme=NOT_PROVIDED,
...     settings__font__size=20,
... )
>>> user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
>>> print(user_preferences.settings)
{'notifications': True, 'font': {'size': 20, 'name': 'Arial'}}
```

This feature is implemented using the newly added databases functions `JSONSet` and `JSONRemove` which serves as abstraction on top of the suitable functions on each database backend.

This PR is a part of the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2024/projects/dav6kK0h) program.

# Checklist
- [x] This PR targets the `main` branch. 
- [ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/db/backends/oracle/features.py [hunk 1]
```diff
@@ -208,3 +208,7 @@ def supports_aggregation_over_interval_types(self):
     @cached_property
     def bare_select_suffix(self):
         return "" if self.connection.oracle_version >= (23,) else " FROM DUAL"
+
+    @cached_property
+    def supports_partial_json_update(self):
+        return self.connection.oracle_version >= (21,)
```
Comment (laymonage) [inline django/db/backends/oracle/features.py:214]: This seems correct according to https://apex.oracle.com/database-features/

However, the function seems to be documented for 19c as well: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/json_transform.html

According to https://stackoverflow.com/a/75286939 it was backported to 19c in 19.10. Could we update this and see if it works? According to https://code.djangoproject.com/wiki/CI we use v19.3.0.0.0 on the CI though, so that may need to be upgraded, or we can test ourselves.

File: django/db/models/fields/json.py [hunk 2]
```diff
@@ -340,12 +340,16 @@ def __init__(self, key_name, *args, **kwargs):
...
+    def unwrap_transforms(self):
         key_transforms = [self.key_name]
         previous = self.lhs
         while isinstance(previous, KeyTransform):
             key_transforms.insert(0, previous.key_name)
             previous = previous.lhs
+        return previous, key_transforms
+
+    def preprocess_lhs(self, compiler, connection):
+        previous, key_transforms = self.unwrap_transforms()
```
Comment (laymonage) [inline django/db/models/fields/json.py:352]: Could we make a separate commit for this change (i.e. extracting `unwrap_transforms()` from `preprocess_lhs()`) please?

File: django/db/models/fields/json.py [hunk 3]
```diff
@@ -619,6 +636,7 @@ class KeyTransformGte(KeyTransformNumericLookupMixin, lookups.GreaterThanOrEqual
 
 
 class KeyTransformFactory:
+
```
Comment (laymonage) [inline django/db/models/fields/json.py:639]: Please make sure there are no unnecessary whitespace changes.

File: django/db/models/functions/comparison.py [hunk 4]
```diff
@@ -36,6 +37,11 @@ def as_sqlite(self, compiler, connection, **extra_context):
             return super().as_sql(
                 compiler, connection, template=template, **extra_context
             )
+        elif output_type == "JSONField":
+            template = "JSON(%(expressions)s)"
+            return super().as_sql(
+                compiler, connection, template=template, **extra_context
+            )
```
Comment (laymonage) [inline django/db/models/functions/comparison.py:44]: Could we extract the changes in `Cast()`, along with the corresponding test, into a separate commit please?

File: django/db/models/sql/subqueries.py [hunk 5]
```diff
@@ -83,13 +84,42 @@ def add_update_values(self, values):
...
+            if transforms:
+                for transform in transforms:
+                    transform_class = field.get_transform(transform)
+                    if transform_class:
+                        lhs = transform_class(lhs)
+                    else:
+                        raise FieldError(
+                            f"{transform} is not a valid Transform "
+                            f"on {field.__class__.__name__}"
+                        )
```
Comment (laymonage) [inline django/db/models/sql/subqueries.py:107]: This logic seems similar to the one that already exists in `Query.try_transform()`. We've discussed this before and encountered a few issues in doing so, but I still wonder if it's possible to reuse that method. I'll do some experiments later.

File: django/db/models/sql/subqueries.py
Comment (laymonage) [inline django/db/models/sql/subqueries.py:1]: Could we extract the changes in this file and the `Transform.get_update_expression()` method in `lookups.py`, along with the tests (albeit only testing for the `NotImplementedError` and `FieldError`) into a separate commit please?

File: django/db/models/functions/json.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
+from django.db import NotSupportedError
+from django.db.models.constants import LOOKUP_SEP
+from django.db.models.expressions import Func, Value
+from django.db.models.fields.json import JSONField, compile_json_path
+from django.db.models.functions import Cast
+
+
+class JSONSet(Func):
```
Comment (laymonage) [inline django/db/models/functions/json.py:8]: Let's create a separate commit for `JSONSet`+tests, and another commit for `JSONRemove`+tests. Then, create a separate commit for adding `KeyTransform.get_update_expression()` and the corresponding tests.

**Review by laymonage — CHANGES_REQUESTED** [review]
Really great work here! I haven't fully reviewed this, but could we squash the commit history, then break it down into smaller commits to make it easier for others to review?

I can see this broken down into at least these commits:

1. Adding `JSONField` handling for `Cast()` on SQLite, and the corresponding tests.
2. Extracting `KeyTransform.unwrap_transforms()` from `KeyTransform.preprocess_lhs()`
3. `JSONSet` + tests
4. `JSONRemove` + tests
5. Changes to `UpdateQuery.add_update_values()` in `subqueries.py` and addition of `Transform.get_update_expression()` in `lookups.py` + tests for the `NotImplementedError` and `FieldError`
6. Addition of `KeyTransform.get_update_expression()` + tests for using key transforms in `.update()`

Thanks!

File: django/db/backends/oracle/features.py (hunk 1 above)
Comment (adzshaf) [reply django/db/backends/oracle/features.py:214]: We agreed on following the https://apex.oracle.com/database-features/ website.

File: django/db/models/fields/json.py (hunk 2 above)
Comment (adzshaf) [reply django/db/models/fields/json.py:352]: Done :+1:

File: django/db/models/functions/comparison.py (hunk 4 above)
Comment (adzshaf) [reply django/db/models/functions/comparison.py:44]: Done :heavy_check_mark:

File: django/db/models/sql/subqueries.py
Comment (adzshaf) [reply django/db/models/sql/subqueries.py:1]: Done :heavy_check_mark:

File: django/db/models/functions/json.py (hunk 6 above)
Comment (adzshaf) [reply django/db/models/functions/json.py:8]: Done :heavy_check_mark:

File: tests/db_functions/json/test_json_set.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+from django.db.models import F, JSONField, Value
+from django.db.models.functions import JSONObject
+from django.db.models.functions.json import JSONSet
+from django.test import TestCase, skipIfDBFeature, skipUnlessDBFeature
+
+from ..models import UserPreferences
+
+
+@skipUnlessDBFeature("supports_partial_json_update")
+class JSONSetTests(TestCase):
```
Comment (laymonage) [inline tests/db_functions/json/test_json_set.py:14]: I think many of the test methods here can be combined into a single method with `subTest`.

File: tests/db_functions/json/test_json_set.py [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+            {"colors": ["black", "yellow", "red", "green"], "notifications": True},
+        )
+
+        # Update using a nonexistent index that is greater than the array length.
+        UserPreferences.objects.update(settings=JSONSet("settings", colors__10="white"))
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+
+        # It should append the value to the array if the database supports it.
+        if not connection.features.json_set_array_append_requires_length_as_index:
+            if connection.vendor == "oracle":
```
Comment (laymonage) [inline tests/db_functions/json/test_json_set.py:245]: What happens here is that we're testing the behaviour of `JSON_SET` when updating an array with an index that's out-of-range. For example, with JSON `{"items": ["a","b","c"]}`.

If we do `JSONSet(field, items__3='d')`, this results in `{"items": ["a","b","c","d"]}` on all databases.

If we do `JSONSet(field, items__4='d')`, however, this gives different results:
- On SQLite, `{"items": ["a","b","c"]}`. The append does not happen.
- On Oracle, `{"items": ["a","b","c",null,"d"]}`. It fills out the "gap" with `null`, before appending `"d"` to the correct index.
- On other databases, `{"items": ["a","b","c","d"]}`. It just appends to the last index of the array.

We weren't sure if this is worth adding another feature flag, so we kind of combined Oracle and SQLite's behaviour into a single `json_set_array_append_requires_length_as_index` flag, with the base assumption being that the append does not happen, and Oracle being the outlier.

Perhaps, we should only do the full array assertion inside a vendor check for both `oracle` and `sqlite`. Then, add a general assertion that `field["items"][3]` does not equal `"d"` – this is true for both Oracle and SQLite, and we still leave room for external DB backends which may have the flag also set to true but we don't know the exact behaviour.

**Review by laymonage — COMMENTED** [review]
Thanks!

@adzshaf and I have been iterating on this together for quite a while. I have yet to fully review the tests and docs, but in general we're pretty happy with the implementation's approach.

There are bits I'm still not sure about, mostly around whether the update expression for e.g. `.update(json_field__key2__key2a=123)` should be implemented in the `KeyTransform` class vs. the `JSONField` itself. There are implications toward future extensibility for other transforms, so I'd like to get the community's opinions on this.

I've left some comments to add a bit of context that hopefully helps explain the intentions.

**Comment by laymonage** [comment]
@bmispelon :shipit:

File: docs/topics/db/queries.txt [hunk 9]
```diff
@@ -1332,6 +1332,199 @@ For example:
     >>> Dog.objects.filter(data__has_any_keys=["owner", "breed"])
     <QuerySet [<Dog: Rufus>, <Dog: Meg>]>
 
+.. _updating-jsonfield:
+
+Updating ``JSONField``
+======================
+
```
Comment (sarahboyce) [inline docs/topics/db/queries.txt:1339]:
```suggestion

.. versionadded:: 5.2

```

File: docs/ref/models/database-functions.txt [hunk 10]
```diff
@@ -1982,3 +1982,63 @@ gaps.
...
+    >>> UserPreferences.objects.update(settings=JSONSet("settings", theme="light"))
+    1
+    >>> user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+    >>> print(user_preferences.settings)
+    {'theme': 'light', 'notifications': True}
+
+
+``JSONRemove``
+--------------
+
```
Comment (sarahboyce) [inline docs/ref/models/database-functions.txt:2022]:
```suggestion

.. versionadded:: 5.2

```

Also think it's worth writing a release note for JSONRemove in the minor features as part of b90bf8f547c4ddcf3d84abf9e1dc899bdc039618

File: docs/ref/models/database-functions.txt [hunk 11]
```diff
@@ -1982,3 +1982,63 @@ gaps.
...
+======================
+
+We'll be using the following model in examples of each function::
+
+    class UserPreferences(models.Model):
+        settings = models.JSONField(encoder=DjangoJSONEncoder)
+
+``JSONSet``
+-----------
+
```
Comment (sarahboyce) [inline docs/ref/models/database-functions.txt:1998]:
```suggestion

.. versionadded:: 5.2

```

I also think it might be worth adding a release not for `JSONSet` in the "Minor features" section as part of this commit (3600f05c99bdb767b17da1715175ef184ced8e75)

File: tests/update/tests.py [hunk 12]
```diff
@@ -346,3 +358,199 @@ def test_order_by_update_on_related_field(self):
...
+        UserPreferences.objects.update(settings__theme__color="white")
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+        self.assertEqual(
+            user_preferences.settings,
+            {
+                "theme": {"color": "white", "font": "Arial"},
+                "notifications": {"email": False, "sms": True},
+            },
+        )
+
```
Comment (sarahboyce) [inline tests/update/tests.py:409]: So I was curious what happens when we do `update_or_create` using this syntax and the answer is, it doesn't fail but it doesn't update (the following test fails expecting it would update).
```suggestion

    def test_update_or_create_json_key_transform(self):
        user_preferences = UserPreferences.objects.create(
            settings={
                "theme": {"color": "black", "font": "Arial"},
                "notifications": {"email": False, "sms": True},
            }
        )
        updated_user_preferences, created = UserPreferences.objects.update_or_create(
            defaults={"settings__theme__color": "white"},
            id=user_preferences.id
        )
        self.assertEqual(created, False)
        self.assertEqual(
            updated_user_preferences.settings,
            {
                "theme": {"color": "white", "font": "Arial"},
                "notifications": {"email": False, "sms": True},
            },
        )

```

Can you investigate whether adding support for `update_or_create` feasible?

**Review by sarahboyce — COMMENTED** [review]
Thank you @adzshaf :rocket: this looks really strong

I had a good catch up with @laymonage so I am a little up to speed

I have added a couple of notes :+1:

File: tests/update/tests.py (hunk 12 above)
Comment (laymonage) [reply tests/update/tests.py:409]: Thanks Sarah, that's a good point! We haven't investigated into adding transforms support for `update_or_create`, but thanks to your review we looked into how `JSONSet` and `JSONRemove` behaves with `update_or_create()`.

We're going to open a separate PR for adding those two functions. In the meantime here are our findings for using `update_or_create` with the functions.

With the following example:

```py
updated_user_preferences, created = UserPreferences.objects.update_or_create(
    defaults={"settings": JSONSet("settings", theme__color="white")},
    id=user_preferences.id,
)
```

- In the case where it does an update (`created is False`), it worked well – the test passed.
- In the case where it does an insert (`created is True`), it fails because we'd be referring to the column for the row which hasn't been inserted yet. It makes sense that this would fail, but I think it would be nice if `JSONSet()` and `JSONRemove()` can fall back to an empty JSON object as the starting point in the insert case.
  - The main blocker we found in this case is that the `for_save` parameter in [`BaseExpression.resolve_expression()`](https://docs.djangoproject.com/en/stable/ref/models/expressions/#django.db.models.Expression.resolve_expression) doesn't let you differentiate whether it's an `INSERT` or an `UPDATE` query – it's set to `True` in both cases.
  - We hacked the places where we pass `for_save=True` and changed it to pass the strings `"create"` and `"update"` depending on the query, and we were able to make `JSONSet()` and `JSONRemove()` resolve the expression accordingly. See https://github.com/adzshaf/django/commit/207598f41e845b9f1db8dcdb3dc1f2850d009eeb for more details.
  - I'm not sure the best course of action on that, though. That being said, I just noticed [this forum thread](https://forum.djangoproject.com/t/add-support-for-atomic-upserts/35813/5) about atomic upserts that would also benefit from the ability to differentiate between `INSERT` vs `UPDATE` in `resolve_expression()`. I'll ask there and see if others have an opinion.

**Review by laymonage — COMMENTED** [review]
Just a quick update: @adzshaf and I still work on this regularly every week. We're going to split this PR into two:
- Adding `JSONSet()` and `JSONRemove()`
- Adding support for using key transforms in `QuerySet.update()` for `JSONFields` (which uses `JSONSet` and `JSONRemove` under the hood)

Along with a few more tweaks based on @sarahboyce's feedback at the DjangoCon US sprints. We found a few more edge cases so it's taking us quite a bit of time, but hopefully we'll open a new PR soon.

File: django/db/models/fields/json.py [hunk 13]
```diff
@@ -390,6 +394,19 @@ def as_sqlite(self, compiler, connection):
...
 
+    def get_update_expression(self, value, lhs=None):
+        from ..functions import JSONRemove, JSONSet
+
+        field, key_transforms = self.unwrap_transforms()
+
+        if lhs is None:
+            lhs = field
+
+        if value is NOT_PROVIDED:
```
Comment (charettes) [inline django/db/models/fields/json.py:405]: I like the idea of using a sentinel to differentiate between `null` and key deletions ([it's something that I believe is missing](https://code.djangoproject.com/ticket/35381#comment:3) for JSON `null` vs  SQL `NULL`) bet I'm not sure we should piggy back on that one since it was meant for a completely different purpose.

**Review by charettes — COMMENTED** [review]
Following the ping from [this forum thread](https://forum.djangoproject.com/t/add-support-for-atomic-upserts/35813/10) from @laymonage 

I think the idea of making `for_save` a `(enum.Enum, str)` to differentiate between different scenario could could as most usage out there are doing `if for_save` and most likely don't need to differentite between both.

My immediate concerns with these changes are that they are possibly trying to do too much at once?

Is there a reason why we can't first land support for `JSONSet` / `JSONRemove` in a first set of changes targeted for 5.2, let them simmer for a release, and then consider adding support for the transform combining feature later on?

I like the ideas put forward by this patch and the separation of concern between field and transform logic but these are more fundamental changes to the ORM works and ask larger questions. For example, should we support `save(update_fields)` with transforms as well? Is it time to introduce a specialized type of transform that denotes nested data structure access instead of trying to push the current abstraction in this space?

IMO these questions can likely be more easily answered while users are allowed to play with `JSONSet` and `JSONRemove` and these primitives are iterated on to iron out bugs first.

File: django/db/models/fields/json.py (hunk 13 above)
Comment (laymonage) [reply django/db/models/fields/json.py:405]: I agree @charettes! This was a temporary solution as suggested by Markus the other day, and I linked to your exact comment: https://fosstodon.org/@laymonage/112949402890177435 (Like I said, big fan 😂)

> I think the idea of making `for_save` a `(enum.Enum, str)` to differentiate between different scenario could could as most usage out there are doing `if for_save` and most likely don't need to differentite between both.

That's what I thought too – though I suppose it could still break if people check for the boolean explicitly e.g. `if for_save is True`. Still doable, just need to be more careful when we do the transition.

> My immediate concerns with these changes are that they are possibly trying to do too much at once?
> 
> Is there a reason why we can't first land support for `JSONSet` / `JSONRemove` in a first set of changes targeted for 5.2, let them simmer for a release, and then consider adding support for the transform combining feature later on?

No particular reason, really. When we implemented the first pass for the `JSONSet` and `JSONRemove` functions, we were curious how much work it would take to implement the transforms support in `update()` directly. Turns out it wasn't much, but indeed it's a bit more "controversial".

Sarah and I discussed this at the DjangoCon US sprints and we decided to split this PR into two. One for the DB functions, and one for the nicer syntactic sugar via `update()` directly – just like you said. I noted this in my comment https://github.com/django/django/pull/18489#pullrequestreview-2395014247 just yesterday.

@adzshaf and I are still working on polishing the DB functions to be ready for a new, smaller PR. We found a few more edge cases so it's taking us a bit longer than we thought. We'll post an update on the ticket when we put the PR up 😄 

> I like the ideas put forward by this patch and the separation of concern between field and transform logic but these are more fundamental changes to the ORM works and bring larger questions. For example, should we support `save(update_fields)` with transforms as well? Is it time to introduce a specialized type of transform that denotes nested data structure access instead of trying to push the current abstraction in this space?
> 
> IMO these questions can likely be more easily answered while users are allowed to play with `JSONSet` and `JSONRemove` and these primitives are iterated on to iron out bugs first.

Thank you, and yes there are definitely more areas of the ORM that we haven't considered like `bulk_{create,update}`, `update_or_create`, etc. and `save(update_fields)` is one of them. I'd be happy to focus on getting `JSONSet` and `JSONRemove` in first, then discuss the more uncertain bits after.

Comment (charettes) [reply django/db/models/fields/json.py:405]: Thanks for the update, it's great to see you push all of these JSONField improvement forward. It's has been a game changer to be able to use `JSONField` for non-Postgres database and a lot of that is thanks to you 🙇 

I think that the two-commit approach has a lot of merit considering how hard it is to get these things right the first time. Let me know when things have shappen if you'd like more feedback. As I said already I think the proposed approach is great in the sense that it demonstrates that it can be done without too many invasive changes.

**Comment by adzshaf** [comment]
Just like what @laymonage mentioned, I have created separate PR for adding `JSONSet` and `JSONRemove`:
- https://github.com/django/django/pull/18758

Once that PR is merged, we will revisit the support for using key and path transforms in `update()` in a new PR.

Thank you @sarahboyce @charettes for the reviews :heart:
